### PR TITLE
add protocol balances for umami finance

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #(issue_number)

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+                    if balance == 0:
+                        continue
+
+                    gm_vault_token_address = self.get_gm_vault_address_by_index(idx)
+                    gm_vault_token = GlobalDBHandler().get_evm_token(
+                        address=gm_vault_token_address,
+                        chain_id=self.evm_inquirer.chain_id,
+                    )
+                    assert gm_vault_token is not None, f'vault token {gm_vault_token_address} must be present'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:173]: Again you can't use assertions here. Please understand what an assertion means.

It means a completely invalid program state. It's something internal. Hitting an assertion means **a programming error**. It should **never happen**

You are asserting on external data. Never do that. These should be checked with if and proper RemoteError or DeserializationError or log.error and continue be done depending on the situation

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+                    underlying_vault_token = gm_vault_token.underlying_tokens[0]
+                    underlying_token_price = Inquirer().find_usd_price(Asset(underlying_vault_token.get_identifier(self.evm_inquirer.chain_id)))  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:176]: same comment for this price being zero here as above.

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                    amount = token_normalized_value_decimals(
+                        token_amount=balance,
+                        token_decimals=gm_vault_token.decimals,
+                    )
+                    price_per_share = self.evm_inquirer.call_contract(
+                        contract_address=gm_vault_token.evm_address,
+                        abi=UMAMI_ASSET_VAULT_ABI,
+                        method_name='pps',
+                    )
+                    formatted_price_per_share = token_normalized_value_decimals(
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:186]: the code here is very similar to the code above. Abstract them into a common fucntion and use them once for staking and once for deposit balances.

**Review by LefterisJP — COMMENTED** [review]
Lots of comments. Also this does not work for me. Exception raised:

```
[05/11/2024 20:49:05 CET] ERROR rotkehlchen.api.rest Greenlet with id 134197913740800: Greenlet for task 28 dies with exception: Unexpected umami vault index 1.
Exception Name: <class 'AssertionError'>
Exception Info: Unexpected umami vault index 1
Traceback:
   File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/api/rest.py", line 455, in _do_query_async
    result = command(self, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/api/rest.py", line 820, in query_blockchain_balances
    balances = self.rotkehlchen.chains_aggregator.query_balances(
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/utils/mixins/lockable.py", line 48, in wrapper
    result = f(wrappingobj, *args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/utils/mixins/cacheable.py", line 123, in wrapper
    result = f(wrappingobj, *args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/aggregator.py", line 561, in query_balances
    getattr(self, query_method)(ignore_cache=ignore_cache)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/utils/mixins/lockable.py", line 48, in wrapper
    result = f(wrappingobj, *args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/utils/mixins/cacheable.py", line 123, in wrapper
    result = f(wrappingobj, *args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/aggregator.py", line 1054, in query_arbitrum_one_balances
    self._query_protocols_with_balance(chain_id=ChainID.ARBITRUM_ONE)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/aggregator.py", line 1132, in _query_protocols_with_balance
    protocol_balances = protocol_with_balance.query_balances()
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/arbitrum_one/modules/umami/balances.py", line 168, in query_balances
    gm_vault_token_address = self.get_gm_vault_address_by_index(idx)
    ^^^^^^^^^^^^^^^^^
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/arbitrum_one/modules/umami/balances.py", line 221, in get_gm_vault_address_by_index
    raise AssertionError(f'Unexpected umami vault index {idx}')
```

File: rotkehlchen/chain/arbitrum_one/modules/umami/decoder.py [hunk 4]
```diff
@@ -75,6 +76,7 @@ def _decode_deposit_withdraw_request(
             event.event_subtype = event_subtype
             event.notes = event_notes
             event.counterparty = CPT_UMAMI
+            event.product = EvmProduct.STAKING
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/arbitrum_one/modules/umami/decoder.py:79]: this is not required. you can just use `addresses_with_activity`, set type/subtype instead and skip product (let it `None`). I mean its' not inherently wrong but wonder why you needed this change.

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        Balances are converted to underlying token amounts using the vault's PPS (price per share).
+        """
+        balances: BalancesSheetType = defaultdict(BalanceSheet)
+        if len(addresses_with_deposits := list(self.addresses_with_deposits(products=[EvmProduct.STAKING]))) == 0:  # noqa: E501
+            return balances
+
+        chunk_size, call_order = get_chunk_size_call_order(self.evm_inquirer)
+
+        # get unstaked balances
+        for vault_token in self.vault_tokens:
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:78]: > Or will the ones in the global DB be only the ones the user interacted with?

yeah. that's all the decoder uses.

File: rotkehlchen/types.py [hunk 6]
```diff
@@ -97,6 +97,7 @@
 SPAM_PROTOCOL = 'spam'
 GEARBOX_PROTOCOL = 'gearbox'
 HOP_PROTOCOL_LP = 'hop_lp'
+UMAMI_PROTOCOL: Literal['umami'] = 'umami'
```
Comment (LefterisJP) [inline rotkehlchen/types.py:100]: you are mixing counterparty and protocol here. Undo this and use `CPT_UMAMI` everywhere. No need to introduce a longer name. If it's also used as a protocol that's fine.

And all constants should be marked with `Final`

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+class UmamiBalances(ProtocolWithBalance):
+    def __init__(
+            self,
+            evm_inquirer: 'ArbitrumOneInquirer',
+            tx_decoder: 'ArbitrumOneTransactionDecoder',
+    ):
+        super().__init__(
+            tx_decoder=tx_decoder,
+            counterparty=UMAMI_PROTOCOL,
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:43]: as I said undo this change you did everywhere.

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        )
+        gm_vault_addresses = [
+            string_to_evm_address('0x959f3807f0Aa7921E18c78B00B2819ba91E52FeF'),
+            string_to_evm_address('0x4bCA8D73561aaEee2D3a584b9F4665310de1dD69'),
+            string_to_evm_address('0xcd8011AaB161A75058eAb24e0965BAb0b918aF29'),
+            string_to_evm_address('0x5f851F67D24419982EcD7b7765deFD64fBb50a97'),
+        ]
+        chunk_size, call_order = get_chunk_size_call_order(self.evm_inquirer)
+        # get unstaked balances
+        for vault_token in vault_tokens:
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:88]:
```suggestion
        for vault_token in vault_tokens:  # get unstaked balances
```

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+                        balances=balances,
+                        vault_token=vault_token,
+                        user_address=user_address,
+                    )
+            except RemoteError as e:
+                log.error(f'Failed to query umami balances for vault {vault_token.evm_address}: {e!s}')  # noqa: E501
+                continue
+
+        # get staked balances.
+        for user_address in addresses_with_deposits:
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:134]:
```suggestion
        for user_address in addresses_with_deposits:  # get staked balances
```

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+                    if balance == 0:
+                        continue
+
+                    gm_vault_token_address = gm_vault_addresses[idx]
+                    gm_vault_token = GlobalDBHandler().get_evm_token(
+                        address=gm_vault_token_address,
+                        chain_id=self.evm_inquirer.chain_id,
+                    )
+                    if gm_vault_token is None:
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:168]: start using the `:=` operator

```suggestion
                    if (gm_vault_token := GlobalDBHandler().get_evm_token(
                        address=gm_vault_token_address,
                        chain_id=self.evm_inquirer.chain_id,
                    )) is None:
```

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+            self,
+            balance: int,
+            vault_token: 'EvmToken',
+            balances: 'BalancesSheetType',
+            user_address: 'ChecksumEvmAddress',
+    ) -> None:
+        """Process vault balance and add to user's balance sheet.
+
+        Args:
+            vault_token: The vault token
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:194]: keep arguments in docstring same order as arguments in function please. But do you even need succh a docstring? This is just reading out the name of the variable. The point of good argument naming and typing is to not need these kind of docstrings

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            balance: Raw balance amount
+            user_address: Address of the user
+            balances: Balance sheet to update
+        """
+        price = get_umami_vault_token_price(
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:199]: `:=` operator

File: rotkehlchen/chain/arbitrum_one/modules/umami/utils.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        inquirer: 'Inquirer',
+        vault_token: 'EvmToken',
+        evm_inquirer: 'EvmNodeInquirer',
+) -> Price | None:
+    """Gets token price for Umami vault.
+
+    Multiplies vault's price per share by the underlying token's USD price.
+
+    Args:
+        inquirer: For underlying token price lookup.
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/utils.py:31]: again same thing here. Really no need for such huge docstrings with the arguments of the function

File: rotkehlchen/chain/arbitrum_one/modules/umami/utils.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        Price: USD price of underlying token, None if pricing fails
+    """
+    try:
+        price_per_share = evm_inquirer.call_contract(
+            contract_address=vault_token.evm_address,
+            abi=UMAMI_ASSET_VAULT_ABI,
+            method_name='pps',
+        )
+    except RemoteError as e:
+        log.error(f'Failed to get price per share for vault {vault_token.evm_address}: {e}')
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/utils.py:45]:
```suggestion
        log.error(f'Failed to get price per share for umami vault {vault_token.evm_address}: {e}')
```

File: rotkehlchen/chain/arbitrum_one/modules/umami/utils.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        return None
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/utils.py:46]: why not return zero here too? In the places you call this you end up having to do an extra check for `None` and one for `ZERO`.

you can log the error and return zero price.

File: rotkehlchen/chain/arbitrum_one/modules/umami/utils.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+    formatted_pps = token_normalized_value_decimals(
+        token_amount=price_per_share,
+        token_decimals=vault_token.decimals,
+    )
+    underlying_token = vault_token.underlying_tokens[0]
+    underlying_price = inquirer.find_usd_price(Asset(underlying_token.get_identifier(evm_inquirer.chain_id)))  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/utils.py:53]: no need for a variable here

Also shouldn't you check that the vault has an udnerlying token in the first place and error log and return zero if it does not? Or is tht checked somewhere else?

```suggestion
    underlying_price = inquirer.find_usd_price(Asset(underlying_token[0].get_identifier(evm_inquirer.chain_id)))  # noqa: E501
```

File: rotkehlchen/inquirer.py [hunk 17]
```diff
@@ -1267,6 +1271,23 @@ def find_hop_lp_price(self, lp_token: EvmToken) -> Price | None:
             log.error(f'Failed to query virtual price of Hop lp token {lp_token.evm_address!s} in {evm_manager.node_inquirer.chain_name!s}. {e!s}')  # noqa: E501
             return ZERO_PRICE
 
+    def find_umami_vault_price(self, token: EvmToken) -> Price | None:
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:1274]: what's the point ofthis function? Can't you just call `get_umami_vault_token_price(` directly in the one place you use it? Please don't do useless things and spend effort for nothing. We are wasting time.

File: rotkehlchen/chain/arbitrum_one/modules/umami/balances.py [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+                        method_name='userInfo',
+                        arguments=[idx, user_address],
+                    )
+
+                    if balance == 0:
+                        continue
+
+                    gm_vault_token_address = gm_vault_addresses[idx]
+                    if (gm_vault_token := GlobalDBHandler().get_evm_token(
+                        address=gm_vault_token_address,
```
Comment (LefterisJP) [inline rotkehlchen/chain/arbitrum_one/modules/umami/balances.py:163]: Can even do better

```suggestion                    
                    if (gm_vault_token := GlobalDBHandler().get_evm_token(
                        address=(gm_vault_token_address := gm_vault_addresses[idx]),
```

**Review by LefterisJP — APPROVED** [review]
Small nitpick  but if all good we can do in another PR if at all

Need to run tests now and see they good but I also need some sleep

**Comment by LefterisJP** [comment]
only `rotkehlchen/tests/unit/globaldb/test_globaldb_consistency.py::test_asset_updates_consistency_with_packaged_db` failed which was expected
